File: Python/day_8.py
def solve():

    with open("../inputs/day_8.txt", "r") as f:
        lines = f.readlines()

    trees = [[] for i in range(len(lines[-1]))]
    visible = [[False for i in range(len(lines[-1]))] for j in range(len(lines[-1]))]

    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if j < len(lines[-1]):
                trees[i].append(int(char))

    # visible from the left
    for row in range(len(trees)):
        start_height = trees[row][0]
        visible[row][0] = True
        for col in range(1, len(trees[0])):
            if trees[row][col] > start_height:
                visible[row][col] = True
                start_height = trees[row][col]

    # visible from the right
    for row in range(len(trees) - 1, -1, -1):
        start_height = trees[row][-1]
        visible[row][-1] = True
        for col in range(len(trees[0]) - 2, -1, -1):
            if trees[row][col] > start_height:
                visible[row][col] = True
                start_height = trees[row][col]

    # visible from the top
    for col in range(len(trees[0])):
        start_height = trees[0][col]
        visible[0][col] = True
        for row in range(1, len(trees)):
            if trees[row][col] > start_height:
                visible[row][col] = True
                start_height = trees[row][col]

    # visible from the bottom
    for col in range(len(trees[0]) - 1, -1, -1):
        start_height = trees[-1][col]
        visible[-1][col] = True
        for row in range(len(trees) - 2, -1, -1):
            if trees[row][col] > start_height:
                visible[row][col] = True
                start_height = trees[row][col]

    visible_count = 0
    for row in visible:
        for char in row:
            visible_count += 1 if char is True else 0

    max_view_score = 1
    for row in range(1, len(trees) - 1):
        for col in range(1, len(trees[0]) - 1):
            curr_score = 1
            curr_val = trees[row][col]

            # up
            i = row - 1
            score_up = 0
            while i >= 0:
                if trees[i][col] < curr_val:
                    score_up += 1
                else:
                    score_up += 1
                    break
                i -= 1
            curr_score *= score_up

            # down
            i = row + 1
            score_down = 0
            while i < len(trees):
                if trees[i][col] < curr_val:
                    score_down += 1
                else:
                    score_down += 1
                    break
                i += 1
            curr_score *= score_down

            # right
            i = col + 1
            score_right = 0
            while i < len(trees):
                if trees[row][i] < curr_val:
                    score_right += 1
                else:
                    score_right += 1
                    break
                i += 1
            curr_score *= score_right

            # left
            i = col - 1
            score_left = 0
            while i >= 0:
                if trees[row][i] < curr_val:
                    score_left += 1
                else:
                    score_left += 1
                    break
                i -= 1
            curr_score *= score_left

            max_view_score = max(curr_score, max_view_score)

    print(f"Part A: {visible_count}")
    print(f"Part B: {max_view_score}")


if __name__ == "__main__":
    solve()

File: Rust/src/day_8.rs

pub fn solution() {
    let input: Vec<Vec<u8>> = include_str!("../../inputs/day_8.txt")
        .split("\n")
        .map(|n| n.as_bytes().iter().map(|&b| b - 48).collect())
        .collect();

    let mut visible: Vec<Vec<bool>> = vec![vec![false; input.len()]; input[0].len()];
    // to right
    for row in 0..input.len() {
        let mut curr_height = input[row][0];
        visible[row][0] = true;
        for col in 1..input[0].len() {
            if input[row][col] > curr_height {
                visible[row][col] = true;
                curr_height = input[row][col];
            }
        }
    }
    // to left
    for row in (0..input.len()).rev() {
        let mut curr_height = input[row][input.len() - 1];
        visible[row][input.len() - 1] = true;
        for col in (0..(input.len() - 1)).rev() {
            if input[row][col] > curr_height {
                visible[row][col] = true;
                curr_height = input[row][col];
            }
        }
    }
    // going up
    for col in 0..input[0].len() {
        let mut curr_height = input[input.len() - 1][col];
        visible[input.len() - 1][col] = true;
        for row in (0..(input.len() - 1)).rev() {
            if input[row][col] > curr_height {
                visible[row][col] = true;
                curr_height = input[row][col];
            }
        }
    }
    // going down
    for col in 0..input[0].len() {
        let mut curr_height = input[0][col];
        visible[0][col] = true;
        for row in 1..input.len() {
            if input[row][col] > curr_height {
                visible[row][col] = true;
                curr_height = input[row][col];
            }
        }
    }
    let mut total_a: i32 = 0;
    for row in 0..visible.len() {
        for col in 0..visible[0].len() {
            if visible[row][col] == true {total_a += 1;}
        }
    }
    println!("{}", total_a);

    let mut view_score = 0;
    for row in 1..(input.len() - 1) {
        for col in 1..(input[0].len() - 1) {
            let mut curr_score = 1;

            // up
            let mut i: i32 = row as i32 - 1;
            let mut up = 0;
            let mut curr_val = input[row][col];
            while i >= 0 {
                if input[i as usize][col] >= curr_val {
                    up += 1;
                    break;
                } else {up += 1;}
                i -= 1;
            }
            curr_score = curr_score * up;

            // down
            let mut i: i32 = row as i32 + 1;
            let mut down = 0;
            let mut curr_val = input[row][col];
            while i < input.len() as i32 {
                if input[i as usize][col] >= curr_val {
                    down += 1;
                    break;
                } else {down += 1;}
                i += 1;
            }
            curr_score = curr_score * down;

            // left
            let mut i: i32 = col as i32 - 1;
            let mut left = 0;
            let mut curr_val = input[row][col];
            while i >= 0 {
                if input[row][i as usize] >= curr_val {
                    left += 1;
                    break;
                } else {left += 1;}
                i -= 1;
            } curr_score = curr_score * left;

            // right
            let mut i: i32 = col as i32 + 1;
            let mut right = 0;
            let mut curr_val = input[row][col];
            while i < input.len() as i32 {
                if input[row][i as usize] >= curr_val {
                    right += 1;
                    break;
                } else {right += 1;}
                i += 1;
            } curr_score = curr_score * right;
            view_score = view_score.max(curr_score);
        }
    }
    println!("{}", view_score);
}